Encode/Decode error with custom type and map

When using a custom type and a map in the same struct, the custom type will be `nil` depending on the order.  From the example I have a custom type `Duration` that parses strings into `time.Duration`.   If the map is left out of the config or comes after the custom type it works fine.  If the custom type is after the map the custom type gets placed in the map and the `Duration` is `nil`:

Example:

Types:

```
type Duration struct {                        
        *time.Duration                                                        
}                                                                 
                                                                             
func (d *Duration) MarshalText() ([]byte, error) {                                   
        return []byte(d.Duration.String()), nil
}                                                                  
                                                                
func (d *Duration) UnmarshalText(text []byte) error {
        x, err := time.ParseDuration(string(text))                   
        if err != nil {                                    
                return err                                                    
        }                                                     
        d.Duration = &x                                                                    
        return nil                                                  
} 

type BasicConfig struct {                                             
        User     string                                    
        Version  string                                                                            
        Labels   map[string]string                                
        Duration *Duration                                                                   
} 
```

## Without a map

```
func TestBasicTomlWithoutCustomType(t *testing.T) {
        dur := time.Second * 5 
        data := BasicConfig{                                          
                User:    "example",
                Version: "0.1",
                Duration: &Duration{
                        &dur,               
                },            
        }                   
        var buf bytes.Buffer    
        if err := toml.NewEncoder(&buf).Encode(data); err != nil {
                t.Fatal(err)          
        }         
                                    
        output := string(buf.Bytes())
        fmt.Println(output)
                               
        var cfg BasicConfig
        if _, err := toml.Decode(output, &cfg); err != nil {
                t.Fatal(err)                                      
        }                   
         
        fmt.Printf("%+v\n", cfg)
                                     
        expected := "5s"   
        result := cfg.Duration.Duration.String()
        if result != expected {
                t.Fatalf("expected %s; received %s", expected, result)
        }                   
} 
```
Result:
```
=== RUN   TestBasicTomlWithoutCustomType
User = "example"
Version = "0.1"
Duration = "5s"

{User:example Version:0.1 Labels:map[] Duration:5s}
--- PASS: TestBasicTomlWithoutCustomType (0.00s)
```

## With preceding map
```
func TestBasicTomlCustomType(t *testing.T) {
        dur := time.Second * 5
        data := BasicConfig{    
                User: "example",          
                Labels: map[string]string{
                        "test": "foo",
                },      
                Duration: &Duration{            
                        &dur,  
                },                                                    
                Version: "0.1",
        }
        var buf bytes.Buffer                               
        if err := toml.NewEncoder(&buf).Encode(data); err != nil {
                t.Fatal(err)
        }      
               
        output := string(buf.Bytes())
        fmt.Println(output)                        
                                                
        var cfg BasicConfig      
        if _, err := toml.Decode(output, &cfg); err != nil {
                t.Fatal(err)
        }
        
        fmt.Printf("%+v\n", cfg)
               
        if cfg.Duration == nil {
                t.Fatal("duration is nil")                                
        }                                
                                          
        expected := "5s"
        result := cfg.Duration.Duration.String()
        if result != expected {                       
                t.Fatalf("expected %s; received %s", expected, result)
        }
}
```
Result:
```
=== RUN   TestBasicTomlCustomType
User = "example"
Version = "0.1"

[Labels]
  test = "foo"
Duration = "5s"

{User:example Version:0.1 Labels:map[test:foo Duration:5s] Duration:<nil>}
--- FAIL: TestBasicTomlCustomType (0.00s)
        utils_test.go:376: duration is nil
FAIL
```

This could very well be something I'm doing wrong :)  Does anything look off in my custom type or how I'm marshalling, etc?

--- a/encode.go
+++ b/encode.go
@@ -423,9 +423,21 @@ func tomlTypeOfGo(rv reflect.Value) tomlType {
 		case encoding.TextMarshaler:
 			return tomlString
 		default:
+			// Someone used a pointer receiver: we can make it work for pointer
+			// values.
+			if rv.CanAddr() {
+				_, ok := rv.Addr().Interface().(encoding.TextMarshaler)
+				if ok {
+					return tomlString
+				}
+			}
 			return tomlHash
 		}
 	default:
+		_, ok := rv.Interface().(encoding.TextMarshaler)
+		if ok {
+			return tomlString
+		}
 		encPanic(errors.New("unsupported type: " + rv.Kind().String()))
 		panic("") // Need *some* return value
 	}

--- a/encode_test.go
+++ b/encode_test.go
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"net"
 	"os"
+	"strings"
 	"testing"
 	"time"
 )
@@ -650,6 +651,71 @@ func TestEncodeError(t *testing.T) {
 				t.Errorf("wrong error\nhave: %q\nwant: %q", err, tt.wantErr)
 			}
 		})
+	}
+}
+
+type (
+	sound struct{ S string }
+	food  struct{ F []string }
+	fun   func()
+	cplx  complex128
+)
+
+// This is intentionally wrong (pointer receiver)
+func (s *sound) MarshalText() ([]byte, error) { return []byte(s.S), nil }
+func (f food) MarshalText() ([]byte, error)   { return []byte(strings.Join(f.F, ", ")), nil }
+func (f fun) MarshalText() ([]byte, error)    { return []byte("why would you do this?"), nil }
+func (c cplx) MarshalText() ([]byte, error) {
+	cplx := complex128(c)
+	return []byte(fmt.Sprintf("(%f+%fi)", real(cplx), imag(cplx))), nil
+}
+
+func TestEncodeTextMarshaler(t *testing.T) {
+	x := struct {
+		Name    string
+		Labels  map[string]string
+		Sound   sound
+		Sound2  *sound
+		Food    food
+		Food2   *food
+		Complex cplx
+		Fun     fun
+	}{
+		Name:   "Goblok",
+		Sound:  sound{"miauw"},
+		Sound2: &sound{"miauw"},
+		Labels: map[string]string{
+			"type":  "cat",
+			"color": "black",
+		},
+		Food:    food{[]string{"chicken", "fish"}},
+		Food2:   &food{[]string{"chicken", "fish"}},
+		Complex: complex(42, 666),
+		Fun:     func() { panic("x") },
+	}
+
+	var buf bytes.Buffer
+	if err := NewEncoder(&buf).Encode(x); err != nil {
+		t.Fatal(err)
+	}
+
+	want := `Name = "Goblok"
+Sound2 = "miauw"
+Food = "chicken, fish"
+Food2 = "chicken, fish"
+Complex = "(42.000000+666.000000i)"
+Fun = "why would you do this?"
+
+[Labels]
+  color = "black"
+  type = "cat"
+
+[Sound]
+  S = "miauw"
+`
+
+	if buf.String() != want {
+		t.Error("\n" + buf.String())
 	}
 }
 

